src/librustc_typeck/check/mod.rs

```diff
@@ -5052,7 +5053,29 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
             self.tcx.generics_of(*def_id).has_self
         }).unwrap_or(false);
 
-        let def_id = def.def_id();
+        let mut new_def = def;
+        let (def_id, ty) = match def {
+            Def::SelfCtor(impl_def_id) => {
+                let ty_and_substs = self.impl_self_ty(span, impl_def_id);
+                let def_id = if let Some(adt_def) = ty_and_substs.ty.ty_adt_def() {
+                    let variant = adt_def.non_enum_variant();
+                    new_def = Def::StructCtor(variant.did, variant.ctor_kind);
+                    variant.did
+                } else {
+                    span_bug!(span, "SelfCtor is not refer to adt type")
```

Comment:
This is not a bug, just an error, you can get this in `Alias` in `impl Alias { ... Self ... }` is not an ADT.